src/index/IndexImpl.h

```diff
@@ -347,10 +348,40 @@ class IndexImpl {
   // --------------------------------------------------------------------------
   // TEXT RETRIEVAL
   // --------------------------------------------------------------------------
+  // This struct is used to retrieve text blocks.
   struct TextBlockMetadataAndWordInfo {
-    TextBlockMetaData tbmd_;
-    bool hasToBeFiltered_;
-    IdRange<WordVocabIndex> idRange_;
+    TextBlockMetadataAndWordInfo(
+        const TextBlockMetaData& tbmd,
+        const IdRange<WordVocabIndex>& includingIdRange)
+        : tbmd_{tbmd},
+          hasToBeFiltered_{hasToBeFiltered(includingIdRange)},
+          optIdRange_{
+              hasToBeFiltered_
+                  ? std::optional<IdRange<WordVocabIndex>>{includingIdRange}
+                  : std::nullopt} {}
+    // The TextBlockMetaData has the information on where the blocks boundaries
+    // and internal boundaries are. It is necessary to retrieve either the
+    // context list or entity list of a text block.
+    const TextBlockMetaData tbmd_;
+
+    // Is set to true if the text block contains entries outside of the
+    // requested range
+    const bool hasToBeFiltered_;
+
+    // The id range of the prefix or word used to retrieve the text block(s). It
+    // is only set if hasToBeFiltered_ was determined to be true during
+    // construction.
+    // Note: This range is inclusive so it is [lowerId, upperId],
+    // NOT [lowerId, upperId)
+    const std::optional<IdRange<WordVocabIndex>> optIdRange_;
+
+    // Returns true if the text block contains entries outside of the requested
+    // range
+    bool hasToBeFiltered(
+        const IdRange<WordVocabIndex>& includingIdRange) const {
+      return !(tbmd_._firstWordId >= includingIdRange.first().get() &&
+               tbmd_._lastWordId <= includingIdRange.last().get());
+    }
```

Comment:
This function can be made private and static, and can be called something like `computeHasToBeFiltered`, 
Then completely throw out the `hasToBeFiltered_` member and implement `hasToBeFiltered()` (a member function) as `optIdRange_.has_value()`.